quipucords/scanner/network/processing/installed_products.py

```diff
@@ -0,0 +1,31 @@
+"""Initial processing of the shell output from the installed_products role."""
+
+import logging
+
+from scanner.network.processing import process
+
+logger = logging.getLogger(__name__)
+
+
+class ProcessInstalledProducts(process.Processor):
+    """Process the installed_products fact."""
+
+    KEY = "installed_products"
+
+    @staticmethod
+    def process(output, dependencies=None):
+        """Process installed_product fact output."""
+        products = []
+        installed_products_cmd_output = output.get("stdout_lines", [])
+        for index, line in enumerate(installed_products_cmd_output):
+            if line.strip().startswith("ID:"):
+                name = ""
+                _, id_value = line.strip().split(":", 1)
+                id_value = id_value.strip()
+                if index + 1 < len(installed_products_cmd_output):
+                    name_line = installed_products_cmd_output[index + 1]
+                    if name_line.strip().startswith("Name:"):
+                        _, name = name_line.strip().split(":", 1)
+                        name = name.strip()
+                products.append({"id": id_value, "name": name})
```

Comment:
```suggestion
        installed_products_cmd_output = output.get("stdout", "")
        # since the command is using grep with surrounding context (-C), we expect each
        # result to be delimited by --
        grep_separator = "--"
        for product in installed_products_cmd_output.split(grep_separator):
            product_dict = {}
            for line in product.splitlines():
                if not line.strip():
                    # skip empty lines
                    continue
                key, value = line.strip().split(":", 1)
                if key in ["Name", "ID"]:
                    product_dict[key.lower()] = value.strip()
            if not product_dict.get("id"):
                # considering the command includes grep "ID:", if we don't parse product
                # with at least ID, there's an error on the implementation.
                logger.error(
                    "Unable to parse relevant product information from the following "
                    f" input\n{product}"
                )
                continue
            products.append(product_dict)
```

I propose a dramatic change here for readability but also to protect us from future errors.

the protection from errors part is the logging message. The rest of it is pretty much optional.

(I personally would just throw an exception and explode everything, but that would change network scan ethos )

Reply:
another option, this one respecting "Name" order (is it on stone that will always be after ID? if so, I think it deserves a comment on code, regardless of any of these snippets being adopted or not.)

```suggestion
        installed_products_cmd_output = output.get("stdout", "")
        # since the command is using grep with surrounding context (-C), we expect each
        # result to be delimited by --
        grep_separator = "--"
        for product in installed_products_cmd_output.split(grep_separator):
            product_dict = {"name": ""}
            for line in product.splitlines():
                stripped_line = line.strip()
                if stripped_line.startswith("ID:"):
                    product_dict["id"] = stripped_line.split(":", 1)[1].strip()
                if product_dict.get("id") and stripped_line.startswith("Name:"):
                    product_dict["name"] = stripped_line.split(":", 1)[1].strip()
                    break
            if not product_dict.get("id"):
                # considering the command includes grep "ID:", if we don't parse product
                # with at least ID, there's an error on the implementation.
                logger.error(
                    "Unable to parse relevant product information from the following "
                    f" input\n{product}"
                )
                continue
            products.append(product_dict)
```